galley/pkg/config/source/kube/inmemory/kubesource.go

```diff
@@ -196,24 +199,65 @@ func (s *KubeSource) RemoveContent(name string) {
 func (s *KubeSource) parseContent(r schema.KubeResources, name, yamlText string) ([]kubeResource, error) {
 	var resources []kubeResource
 	var errs error
-	for i, chunk := range kubeyaml.Split([]byte(yamlText)) {
-		chunk = bytes.TrimSpace(chunk)
+
+	reader := strings.NewReader(yamlText)
+	readCloser := ioutil.NopCloser(reader)
+	decoder := yaml.NewDocumentDecoder(readCloser)
+
+	buf := make([]byte, 4096)
+	for {
+		doc, err := readDocument(decoder, buf)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			e := fmt.Errorf("error reading documents in %s: %v", name, err)
+			scope.Source.Warnf("%v - skipping", e)
+			scope.Source.Debugf("Failed to parse yamlText chunk: %v", yamlText)
+			errs = multierror.Append(errs, e)
+			break
+		}
+
+		chunk := bytes.TrimSpace(doc)
 		r, err := s.parseChunk(r, chunk)
 		if err != nil {
-			e := fmt.Errorf("error processing %s[%d]: %v", name, i, err)
+			e := fmt.Errorf("error processing %s: %v", name, err)
 			scope.Source.Warnf("%v - skipping", e)
 			scope.Source.Debugf("Failed to parse yaml chunk: %v", string(chunk))
 			errs = multierror.Append(errs, e)
 			continue
 		}
 		resources = append(resources, r)
 	}
+
 	return resources, errs
 }
 
+// readDocument is a helper for reading documents from yaml.NewDocumentDecoder.
+// yaml.NewDocumentDecoder returns a Reader instance such that every Read call
+// is an entire document; however like all Readers it requires a byte buffer to
+// write to.
+func readDocument(r io.Reader, buf []byte) ([]byte, error) {
+	var doc []byte
+
+	for {
+		num, err := r.Read(buf)
```

Comment:
Can we use ioutil.readall?

Reply:
I don't think so (or at least not without changing more Galley logic about how to handle these files in general).

The `Reader` returned by `yaml.NewDocumentDecoder` returns each document as a 'chunk' (e.g. a single `Read` call): https://godoc.org/k8s.io/apimachinery/pkg/util/yaml#NewDocumentDecoder

This is both convenient and frustrating, as the caller still needs to provide a buffer, and can still run into errors when the buffer size isn't big enough. Hence all the extra logic I had to add that reads 'chunks', and handles `io.ErrShortBuffer` by continuing to read the parsed chunk.

It's possible (likely!) that there's a utility class/function somewhere that handles readers designed to return significant 'chunks', but my understanding of the generic `Reader` interface is that the read size returned is implementation dependent (and usually not significant). If that's true, I'm unlikely to find any utility classes/functions in the core library at least.